89. Gray Code

```diff
@@ -0,0 +1,26 @@
+//Time: O (2 ^ n) Space: O(2 ^ n)
```

Comment:
Give me a detailed analysis of those Big O answers.

Reply:
I think the gray code of n should be at length 2^n.
So we need roughly 2^n space. And the space needed for n-1 is n/2, and for n-2 is n/4. These sums up result in roughly another O(2^n). So in general the space needed is O(2^n).
And for each recursive we need O(2^n) steps to push numbers, which is the major cost of time complexity.